src/iterators.rs

```diff
@@ -30,16 +30,15 @@ impl<'a, T> Iterator for Ancestors<'a, T> {
     type Item = &'a Node<T>;
 
     fn next(&mut self) -> Option<&'a Node<T>> {
-        if let Some(current_id) = self.node_id.clone() {
```

Comment:
I'm not sure why there was a `clone()` here so I removed it from me implementation.

Reply:
Let me start by saying, I love the way you re-wrote this!

However, there's a subtle difference in behavior here.  It may not really matter, but if you can think of a clean way of dealing with this, it'd be nice.

Previously, `self.node_id` would be set to `None` if the current `Node` doesn't have a parent.  This means the next iteration would essentially be a no-op because the very first check `if let Some(current_id) ...` would fail.

With your changes though, `self.node_id` doesn't ever get set to `None` in that case, so every subsequent call to `next()` would still try and do a lookup (`current_node = self.tree.get(...)` + `current_node.parent()`).  This is probably not an issue most of the time as a `for ... in ...` loop will stop on the first `None` that it gets, but if someone were manually calling `next()` this would be less than optimal.  Also it feels slightly dirty to leave `self.node_id` in a "wrong" state to me (although, what the "correct" state is at the end of iteration is probably debatable).

Again, if there's no good way to handle this then I say don't worry about it, however, I wanted to bring it up just in case.

Reply:
ha! `take` is perfect, good thinking!  and yeah I agree, the `&current_id` is better than doing an `as_ref()` imo.